### PR TITLE
Add E2E test suite CI job with code coverage

### LINKED ISSUE #126: Add E2E test suite CI job with code coverage
## Problem

The PAPI integration tests currently run as **standalone demo scripts** (`bucket-membership.js`, `bucket-with-storage.js`, `s3-lifecycle.js`, `drive-lifecycle.js`, `checkpoint-rewards.js`) invoked individually from the `integration-tests` CI job. This has several issues:

1. **No unified test runner** — each script is a separate `just` recipe, making it hard to add/remove workflows or get a summary of pass/fail across all of them.
2. **No code coverage** — we have no visibility into what percentage of our shared application code (`api.js` with 40+ extrinsic wrappers, `common.js` with chain setup + tx utilities) or the Rust provider-node (19 source files: HTTP API, storage, MMR, checkpoint/agreement coordinators, S3/FS APIs, challenge responder) is actually exercised by the E2E tests.
3. **Redundancy** — the standalone scripts duplicate setup logic and are now superseded by the structured E2E suite in `examples/papi/e2e/` (10 workflows, ~96 tests) which covers the same flows and more.

### 1. New `e2e-integration-tests` CI job

Add a dedicated CI job that runs the full E2E test suite (`examples/papi/e2e/runner.js`) against a single inmemory provider. This replaces the individual demo script invocations with a single orchestrated run that:
- Executes all 10 workflow files sequentially
- Collects exit codes and prints a summary table
- Runs in its own CI job parallel to the existing `integration-tests` and `sc-integration-tests` jobs

### 2. JS code coverage via c8

[c8](https://github.com/bcoe/c8) uses V8's built-in code coverage (zero overhead). The E2E runner spawns each workflow as a child process via `execFileSync`; c8 sets `NODE_V8_COVERAGE` which is inherited by child processes, so all 10 workflow scripts' coverage is collected automatically.

**What's measured:**
- `api.js` — 40+ extrinsic wrappers + HTTP helpers
- `common.js` — chain setup, tx submission, utilities

**Configuration** (`.c8rc.json`):
- `"all": true` — reports uncovered lines (not just executed files)
- `"include": ["api.js", "common.js"]` — targets only the shared application modules
- `"exclude": ["e2e/**", "node_modules/**", ...]` — keeps test harness out

### 3. Rust provider-node coverage via LLVM instrument-coverage

The provider binary is built with `-C instrument-coverage` into a separate `target-cov/` directory (avoids rebuilding the runtime's deps). Profraw data is written on process exit — SIGTERM from `pkill` triggers clean shutdown and flush.

**What's measured:**
- All 19 source files in `provider-node/src/` (HTTP API, storage layer, MMR, coordinators, S3/FS APIs, challenge responder)

**Report generation:**
- `llvm-profdata merge` combines profraw files
- `llvm-cov report` prints a text summary to CI logs
- `llvm-cov export --format=lcov` produces an LCOV file uploaded as an artifact

### 4. Cleanup of superseded scripts

The standalone demo scripts that are now covered by the E2E suite are removed:
- `bucket-membership.js`, `bucket-with-storage.js`, `checkpoint-rewards.js`, `drive-lifecycle.js`, `s3-lifecycle.js`
- Corresponding `just` recipes are removed from the justfile
- The `integration-tests` CI job no longer invokes these individual scripts

## Build time impact

- **JS coverage (c8)**: Zero build impact — V8 native coverage is essentially free. ~1s overhead to merge coverage JSONs.
- **Rust coverage build**: Building `storage-provider-node` with `-C instrument-coverage` into `target-cov/` requires a full provider build (~5-8 min), but this runs while the chain starts up (~3-5 min), so net wall-clock impact is minimal.

## Artifacts

Each CI run produces downloadable artifacts:
- `js-coverage-{runtime}` — c8 JSON coverage data for api.js + common.js
- `rust-coverage-{runtime}` — LCOV file for provider-node/src/

### PR DESCRIPTION
**Pull request by mudigal**
## Summary

Add a comprehensive E2E test suite (10 workflows, 95 tests) that exercises the full storage pallet lifecycle against a real chain + provider node, with dual JS/Rust code coverage.

### Chain infrastructure: dev mode instead of zombienet

The E2E job runs `polkadot-omni-node` in **standalone dev mode** (`--dev-block-time 2000`) instead of spawning a full zombienet relay+parachain network. This is a strategic change:

- **No relay chain needed** — blocks are produced and finalized instantly by a single node
- **2-second block times** — each extrinsic finalizes in ~2s vs ~12s with zombienet
- **Deterministic finality** — no relay chain consensus delays; blocks finalize immediately
- **~5x faster E2E runs** — full suite completes in ~8 minutes vs ~40+ with zombienet
- **Simpler CI setup** — single binary, no relay chain binary, no zombienet config

The `start-e2e-chain` composite action generates a chain spec via `chain-spec-builder` and launches `polkadot-omni-node --alice --tmp --dev-block-time 2000`.

### Test execution model

- **Sequential runner** (`e2e/runner.js`): discovers `NN-*.js` files, runs each as a child process with 10-minute timeout
- **Shared chain state**: tests run sequentially on the same chain, so later workflows see state from earlier ones (e.g., test 08 sees agreements created in test 01)
- **Single provider node**: only Alice's provider runs (with agreement + checkpoint coordinators), which means tests needing other providers must manually accept agreements
- **JS coverage via c8** (V8 native): measures `api.js` and `common.js` coverage
- **Rust coverage via LLVM**: provider-node built with `-C instrument-coverage`, profraw merged after graceful SIGTERM

### Test workflows

| # | Workflow | Tests | What it covers |
|---|----------|-------|----------------|
| 01 | Provider Registration | 12 | Register, update settings, add stake, multiaddr, matching, failure cases |
| 02 | Agreement Lifecycle | 12 | Request, accept (auto), reject, withdraw, end (Pay/Burn), failure cases |
| 03 | S3 Bucket & Objects | 11 | Create/delete S3 buckets, put/get/copy/delete objects, upsert, paths |
| 04 | Data Upload & Retrieval | 12 | Small/medium/max chunks, binary data, S3 PUT/GET/HEAD/LIST/DELETE, blake2-256 |
| 05 | Checkpoint & Challenges | 7 | Client/provider checkpoints, off-chain/on-chain challenges, rewards |
| 06 | Bucket Membership ACL | 10 | Add/remove members, promote/demote roles, self-demotion, authorization |
| 07 | Agreement Extensions | 7 | Extend duration, top up bytes, block extensions (per-agreement & global) |
| 08 | Provider Deregistration | 8 | Announce/cancel deregistration, active agreements, re-registration |
| 09 | Drive Lifecycle | 7 | Create/share/unshare/delete drives, authorization |
| 10 | Edge Cases | 9 | Balance accounting, capacity tracking, freeze semantics, data integrity |

### Bug fixes discovered during E2E development

Fixed 8 bugs in test logic and shared utilities:

1. **TDZ in `waitForNextBlock`/`waitForBlock`** (`common.js`): ReplaySubject fired callback synchronously before `const sub` was assigned
2. **Process hang after tests**: unhandled rejection skipped `papi.destroy()`, open WebSocket kept Node alive
3. **Wrong field names in `createBucketWithStorage` callers**: passed `max_capacity`/`max_payment` but pallet expects `max_bytes`/`max_price_per_byte`
4. **Test 6.4 cross-admin demotion**: pallet only allows self-demotion, test had admin demoting another admin
5. **`setExtensionsBlocked` API**: passed extra `provider` param pallet doesn't accept, and signed as client instead of provider
6. **Test 8.4 auto-accept**: relied on Charlie's provider node (not running) — switched to manual `acceptAgreement`
7. **Test 7.1 extend duration**: `extend_agreement` resets `expires_at = current_block + additional_duration` (not additive) — `extDuration` must exceed original duration
8. **Test 2.6 Burn encoding**: `Enum("Burn")` missing required `burn_percent` field for `EndAction::Burn { burn_percent: u8 }`

### Other changes

- Removed superseded standalone demo scripts (`bucket-membership`, `bucket-with-storage`, `checkpoint-rewards`, `drive-lifecycle`, `s3-lifecycle`)
- Extracted CI composite actions: `start-e2e-chain`, `start-zombienet`, `wait-for-parachain`, `wait-for-provider-health`
- Added `just e2e` and `just e2e-single` recipes

## Test plan

- [x] All 10 E2E workflows pass (95 tests total)
- [x] JS coverage text summary appears in CI logs
- [ ] Rust provider coverage text summary appears in CI logs
- [ ] `js-coverage-*` and `rust-coverage-*` artifacts downloadable from Actions run
- [ ] Existing `integration-tests` and `sc-integration-tests` jobs still pass unaffected

Closes #126

### CODE REVIEW

File: .github/workflows/integration-tests.yml [hunk 1]
```diff
@@ -199,10 +190,213 @@ jobs:
...
+              fi
+            fi
+            if [ "$i" -eq 180 ]; then
+              echo "TIMEOUT: parachain did not finalize blocks"
+              exit 1
+            fi
+            sleep 5
+          done
+
+      - name: Start provider (inmemory) and wait for health
```
Comment (bkontur) [inline .github/workflows/integration-tests.yml]: Please change it like here: https://github.com/paritytech/web3-storage/pull/128/files#diff-b4361375d6c8e4322510b05dd051905312f3a9086ba0da1a75f13b6f54267ef0R118

Comment (mudigal) [reply .github/workflows/integration-tests.yml]: I now moved all repeated actions into github actions - good one.

File: examples/papi/api.js
Comment (ilchu) [inline examples/papi/api.js:1]: There's some very similar thin wrappers in `provider.state.ts` in UIs. Should we unify all the helper code throughout E2E / SC examples **and** UI, wdyt?

File: examples/papi/common.js [hunk 2]
```diff
@@ -335,44 +359,30 @@ export async function waitForTransaction(
...
- * and does NOT throw when dispatch fails — only when the tx is invalid (bad
- * signature, low nonce, etc). Without this helper, a failed extrinsic looks
- * indistinguishable from a successful one with no events, and the failure
- * surfaces later as a confusing "Expected exactly 1 X event, got 0".
+ * Resolves at best-block INCLUSION, not finalization. On a local zombienet
+ * parachain, finality lags inclusion by several blocks (~36s vs ~6s per tx);
+ * waiting for finalization on every extrinsic is what pushed the integration
+ * demos past the 30-min CI cap. The demos verify outcomes via the emitted
+ * events, which are already present at inclusion — the only thing given up is
+ * reorg-safety, which is moot on a single-collator local chain.
```
Comment (ilchu) [inline examples/papi/common.js]: Speaking from the UI viewpoint again, this is also a major issue for the duration and maybe even flakiness of those playwright e2es. Also worth a followup?

**Review by ilchu — APPROVED** [review]

File: examples/papi/api.js
Comment (bkontur) [reply examples/papi/api.js:1]: > There's some very similar thin wrappers in `provider.state.ts` in UIs. Should we unify all the helper code throughout E2E / SC examples **and** UI, wdyt?

Yes, I think it is a good idea, @danielbui12 created issue for TS SDK, maybe this could be a good place.

Also I want to unify the `ui-e2e.yml`, when #129 and #130 are working

Comment (mudigal) [reply examples/papi/api.js:1]: Makes sense, created an issue for that - https://github.com/paritytech/web3-storage/issues/135. Want to keep this PR for E2E Tests and Coverage.

File: examples/papi/common.js (hunk 2 above)
Comment (mudigal) [reply examples/papi/common.js]: Yes absolutely - https://github.com/paritytech/web3-storage/issues/136 to be followed up.

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -98,86 +159,74 @@ jobs:
...
-          nohup env PROJECT_ROOT=$(pwd) .bin/zombienet spawn -p native ${{ matrix.runtime.zombienet_config }} > /tmp/zombienet.log 2>&1 &
-          ZOMBIE_PID=$!
-          echo "Waiting for zombienet to set up nodes..."
-          sleep 15
-          if ! kill -0 $ZOMBIE_PID 2>/dev/null; then
-            echo "ERROR: zombienet exited early"
-            cat /tmp/zombienet.log || true
-            exit 1
-          fi
+        uses: ./.github/actions/start-zombienet
```
Comment (bkontur) [inline .github/workflows/integration-tests.yml:165]: Let's use here fór sc-integration-tests Also start-e2e-chain

**Review by bkontur — APPROVED** [review]

File: .github/workflows/integration-tests.yml (hunk 3 above)
Comment (mudigal) [reply .github/workflows/integration-tests.yml:165]: already done. Both `sc-integration-tests` and `e2e-integration-tests` use composite actions (`start-zombienet` and `start-e2e-chain` respectively).
